dapp/src/components/AccountListener.js

```diff
@@ -17,98 +18,136 @@ const AccountListener = (props) => {
-      if (!account) return
-
-      try {
-        const [
-          usdtAllowance,
-          daiAllowance,
-          usdcAllowance,
-          ousdAllowance,
-        ] = await Promise.all([
-          displayCurrency(await usdt.allowance(account, vault.address), usdt),
-          displayCurrency(await dai.allowance(account, vault.address), dai),
-          displayCurrency(await usdc.allowance(account, vault.address), usdc),
-          displayCurrency(await ousd.allowance(account, vault.address), ousd),
-        ])
 
+        // allowance
         AccountStore.update((s) => {
           s.allowances = {
-            usdt: usdtAllowance,
-            dai: daiAllowance,
-            usdc: usdcAllowance,
-            ousd: ousdAllowance,
+            ousd: data[4],
+            usdt: data[5],
+            dai: data[6],
+            usdc: data[7],
           }
         })
-      } catch (e) {
+      })
+      .catch((e) =>
         console.error(
-          'AccountListener.js error - can not load account allowances: ',
+          'AccountListener.js error - can not load account balances: ',
           e
         )
-      }
-    }
-    await loadBalances()
-    await loadAllowances()
+      )
   }
 
-  useEffect(() => {
-    if (account) {
-      login(account, setCookie)
+  const subscribeToEvents = (contracts) => {
+    // Polls data first time, then rely on events
+    const { usdt, dai, usdc, ousd, vault } = contracts
+    const rpcProvider = props.rpcProvider
+    const pollNTimes = (n, promiseFn) => {
+      if (n === 0) return
+      // Poll every 5 seconds
+      setTimeout(() => {
+        promiseFn().then(pollNTimes(n - 1, promiseFn))
+      }, 5000)
     }
 
+    const updateOnAllowanceEvent = (contract, name) =>
+      rpcProvider.on(
+        contract.filters.Approval(account, vault.address, null),
+        (result) =>
+          displayCurrency(result.data, contract).then((allowance) =>
+            AccountStore.update((s) => {
+              s.allowances[name] = allowance
+            })
+          )
+      )
+    // Subscribe to Transfer event. Then poll balance once event received
+    const updateOnTransferEvent = (contract, name) => {
+      // Account sends tokens
+      rpcProvider.on(
+        contract.filters.Transfer(account, null, null), // event Transfer(address indexed from, address indexed to, uint tokens);
+        (result) =>
+          pollNTimes(5, () =>
+            contract
+              .balanceOf(account)
+              .then((balance) => displayCurrency(balance, contract))
+              .then((balance) =>
+                AccountStore.update((s) => {
+                  s.balances[name] = balance
+                })
+              )
+          )
+      )
+      // Account receives tokens
+      rpcProvider.on(
+        contract.filters.Transfer(null, account, null), // event Transfer(address indexed from, address indexed to, uint tokens);
+        (result) =>
+          pollNTimes(5, () =>
```

Comment:
`pollNTimes`

For balance updates, we are going off of the Transfer event. This event doesn't tell us the balance, only the amount transferred. So once the code gets the Transfer event, it then polls for the balance. There is a latency when metamask has the updated balance, so the code uses a kludge: using `pollNTimes` to poll metamask several times for an update.